Resiliency: Node manager reconciliation path yields unchecked errors

derailed opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

<!-- Description of change -->

> NOTE: Initial pass!

Found several places in the node manager control loop where we are ignoring potential errors and either not logging them or worth not bubbling up to their respective call sites.

> NOTE! Please go thru this PR with a fine comb as I don't quite have the necessary chops on whether some of these should be hard errors or not. 
Thank you!!

Fixes: #issue-number

Signed-off-by: Fernand Galiana <fernand.galiana@gmail.com>

christarazi commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -281,11 +284,11 @@ func xfrmDeleteConflictingState(states []netlink.XfrmState, new *netlink.XfrmSta
 		}
 	}
-	return deletedSomething, nil
+	return deletedSomething, errs
```
I think what might be a better alternative here is to aggregate the msg and de-dup them a bit because then the error string might get too long given we are at a "leaf" function in the reconciliation loop. 

This can be done by counting the number of failures out of the number of states to delete, i.e. "5/10 failed", and keeping a set of the unique underlying errors that occurred. Given that we are interacting with netlink here, the error returned from those calls are going to be syscall-style errno error types such as EINVAL, ENOMEM, etc.

I would do something like this pseudocode:

```go
var (
	deletedSomething bool
	errorTypes       set.Set[error]
	failed           int
)
...
if err != nil {
	failed++
	errorTypes.Insert(err)
}
...
if failed > 0 {
	return deletedSomething, fmt.Errorf("failed to remove a conflicting XFRM states (%d/%d): %w", failed, len(states), errorTypes.Get()...)
}     
```

derailed replied:
Thanks Chris! Good point!! I'll lasso in something along those lines...

christarazi commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -727,12 +730,14 @@ func DeleteXfrm() {
+	for _, s := range xfrmStateList {
+		if isXfrmStateCilium(s) {
+			if err := netlink.XfrmStateDel(&s); err != nil {
+				log.WithError(err).Warning("deleting old xfrm state failed")
```
This is a good step, but I'd go further: let's make this function return an error and return an aggregated error of how many resources failed to be deleted and a record of the error types observed similar to my above comment. Then we can move the logging of the error to the caller of this function, rather than on each failure to delete an individual resource. I think this makes sense given the role of this function as a "leaf" in terms of XFRM logic and interaction with netlink.

derailed replied:
Hum... I was not sure about this part. The original code just logged and did not propagate. I am guessing because we want to add scope to the logs which will be lost unless we wrap the error further?? But would kill the log indexer if that is the intent here??

christarazi replied:
Yeah, my suggestion would be to try to prevent as much of this pattern where the leaf functions are logging and instead kick the error back up to the caller. Then the caller can decide to log or not. There can be exceptions to this rule depending on the leaf function, but in this case, it seems like it could do with the same thing we implemented in `xfrmDeleteConflictingState()`

derailed replied:
Makes sense. I've changes the call to now return an error.

christarazi commented on pkg/datapath/linux/ipsec/probe_linux.go:
```diff
@@ -72,5 +74,5 @@ func ProbeXfrmStateOutputMask() error {
 		return errors.New("incorrect value for probed IPSec output mask attribute")
 	}
-	return nil
+	return
```
I'm confused by this change. What will be the behavior if the code failed on line 75 and returns the error `incorrect value for probed IPSec output mask attribute` if the deferred function will also execute and return `e`?

derailed replied:
I've tried to keep this code on par with the original intent which would try to delete the state no matter what happens after the defer. You are correct tho the current update would blow the previous err out (if any). I'll fix that.
Not keen on named returned, nor the defer here but thinking there is a combo of hard errors here (ie no state) or `softer` err ie the state is not in the shape we want. I would think the defer needs to move down as there no point in deleting when the state does not exist in the first place?

christarazi replied:
If we are relying on the defer to do something specific and we need to know whether it succeeded or not, then it should not be `defer`red. It should just be a normal function call, because the `defer` just complicates the control flow for no reason.

christarazi commented on pkg/datapath/linux/node.go:
```diff
@@ -566,15 +566,14 @@ func upsertIPsecLog(err error, spec string, loc, rem *net.IPNet, spi uint8) erro
-		metrics.Register(n.ipsecMetricCollector)
+		if err := metrics.Register(n.ipsecMetricCollector); err != nil {
+			log.WithError(err).Errorf("IPSec metrics registration failed. No metrics will be reported!")
+		}
```
This metrics registration code is a bit interesting; I don't think it is possible for it to return errors, **currently**. Looking at the [code](https://github.com/cilium/cilium/blob/d06140fc9e604a39c6cf71ee596cbb0e27691ef3/pkg/metrics/metrics.go#L1432-L1438), the error is never captured. We should probably fix this, cc @dylandreimerink as last author of the code. Once it's captured, then the changes in the PR would actually be utilized. 

I'm also wondering what sort of errors metrics registration could return, but I digress...

For this PR, my only comment is we don't need the dynanmic formatting:

```suggestion
		if err := metrics.Register(n.ipsecMetricCollector); err != nil {
			log.WithError(err).Errorf("IPsec metrics registration failed. No metrics will be reported!")
		}
```

derailed replied:
Right noticed that too. But figure I'll flag it. Since we swallow the error too in the wrapper??

dylandreimerink replied:
Agreed, that just ignoring the errors is not the correct way to deal with that. I don't exactly know why I made that decision. I think I did so because ultimately the goal is to migrate all metrics registration over to using hive instead of these global methods, which would move any error reporting burden to the registry anyway. But until then, we should at least be aware of registration errors.

I think the code is good as is, given the function signature has an error, it should always be checked in case the function starts returning actual errors.

christarazi commented on pkg/datapath/linux/node.go:
```diff
@@ -1828,7 +1830,9 @@ func (n *linuxNodeHandler) NodeConfigurationChanged(newConfig datapath.LocalNode
-	n.updateOrRemoveNodeRoutes(prevConfig.AuxiliaryPrefixes, newConfig.AuxiliaryPrefixes, true)
+	if err := n.updateOrRemoveNodeRoutes(prevConfig.AuxiliaryPrefixes, newConfig.AuxiliaryPrefixes, true); err != nil {
+		return err
+	}
```
Just to provide a bit more context on the error:

```suggestion
	if err := n.updateOrRemoveNodeRoutes(prevConfig.AuxiliaryPrefixes, newConfig.AuxiliaryPrefixes, true); err != nil {
		return fmt.Errorf("failed to update or remove node routes: %w", err)
	}
```

derailed replied:
Right! This is what tracing is for... where does this end if you are in a deep callstack?? net effect errors are a mile long ;( 
This is where I struggled with this pr. At times we add context by wrapping, others we reach out for scoped loggers and hence don't propagate...

christarazi commented on pkg/datapath/linux/node.go:
```diff
@@ -1857,25 +1861,36 @@ func (n *linuxNodeHandler) NodeConfigurationChanged(newConfig datapath.LocalNode
+		}
+		if err := n.updateOrRemoveClusterRoute(n.nodeAddressing.IPv6(), newConfig.EnableIPv6); err != nil {
+			errs = errors.Join(errs, err)
+		}
```
I would also add some context to the errors here as well:

```suggestion
		if err := n.updateOrRemoveClusterRoute(n.nodeAddressing.IPv4(), newConfig.EnableIPv4); err != nil {
			errs = errors.Join(errs, fmt.Errorf("failed to update or remove IPv4 cluster router: %w", err))
		}
		if err := n.updateOrRemoveClusterRoute(n.nodeAddressing.IPv6(), newConfig.EnableIPv6); err != nil {
			errs = errors.Join(errs, fmt.Errorf("failed to update or remove IPv6 cluster router: %w", err))
		}
```

christarazi commented on pkg/datapath/linux/node.go:
```diff
@@ -1857,25 +1861,36 @@ func (n *linuxNodeHandler) NodeConfigurationChanged(newConfig datapath.LocalNode
 		// single cluster route has been disabled, remove route
-		n.deleteNodeRoute(n.nodeAddressing.IPv4().AllocationCIDR(), false)
-		n.deleteNodeRoute(n.nodeAddressing.IPv6().AllocationCIDR(), false)
+		if err := n.deleteNodeRoute(n.nodeAddressing.IPv4().AllocationCIDR(), false); err != nil {
+			errs = errors.Join(errs, err)
+		}
+		if err := n.deleteNodeRoute(n.nodeAddressing.IPv6().AllocationCIDR(), false); err != nil {
+			errs = errors.Join(errs, err)
+		}
```
I would add more context here as well:

```suggestion
		if err := n.deleteNodeRoute(n.nodeAddressing.IPv4().AllocationCIDR(), false); err != nil {
			errs = errors.Join(errs, fmt.Errorf("failed to remove old IPv4 cluster route: %w", err))
		}
		if err := n.deleteNodeRoute(n.nodeAddressing.IPv6().AllocationCIDR(), false); err != nil {
			errs = errors.Join(errs, fmt.Errorf("failed to remove old IPv6 cluster route: %w", err))
		}
```

Additionally, when I think about this code path a bit more, it's a bit worrying if we couldn't delete a previous node route for the cluster when the node changed configuration. Failing to do this could override the desired routing configuration for the node and potentially cause connectivity outages. At the same time, I'm not sure how this could be possible. `UseSingleClusterRoute` comes from the Agent configuration, so if it changes, then a restart is required for it to take effect.

Let's see if @cilium/sig-datapath could explain a bit more about this path and what sort of error visibility makes sense.

pchaigno replied:
Maybe ping the original author? It's unlikely the whole datapath group (28 people) is going to know about this.

christarazi replied:
The original author is Thomas Graf :wink:

pchaigno replied:
Then nobody knows :sweat_smile:

christarazi reviewed (changes requested):
Thanks for the PR!

A few comments to improve the error paths.

gentoo-root commented on pkg/datapath/linux/ipsec/errorset.go:
```diff
@@ -0,0 +1,64 @@
+	}
+}
+
+func (e *setError) add(errs ...error) {
```
This operation is O(n^2). How many items can be in an error set in the worst case?

derailed replied:
Indeed. Great catch! 
I am not sure how often nor how many errors could be reported here. 
Changed the implementation to leverage a map instead.

gentoo-root commented on pkg/datapath/linux/ipsec/errorset.go:
```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package ipsec
```
This structure doesn't look IPsec-specific. Maybe there is a generic location more fitting for it?

derailed replied:
Great point. Thank you Maxim! Was not sure where to park it.

gentoo-root commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -715,27 +717,39 @@ func isXfrmStateCilium(state netlink.XfrmState) bool {
+		if isXfrmPolicyCilium(p) {
+			if err := netlink.XfrmPolicyDel(&p); err != nil {
+				log.WithError(err).Warning("deleting xfrm policy failed")
+				errs = errors.Join(errs, err)
```
Why is setError not used here, unlike the previous function?

derailed replied:
Great observation! Switched the implementation

gentoo-root replied:
Did you? It's still the same.

derailed committed:
Add errorset to track unique errors

Some calls may return multiple non unique errors.
Leveraging an error set in those situations will trim down repetitions
while reporting errors.

Signed-off-by: Fernand Galiana <fernand.galiana@isovalent.com>

derailed committed:
Resiliency: Node manager reconciliation path yields unchecked errors

> NOTE: Initial pass!

Found several places in the node manager control loop where we are ignoring potential
errors and either not logging them or worth not bubbling up to their respective call sites.

> NOTE! Please go thru this PR with a fine comb as I don't have the
necessary understanding on whether some of these should be hard errors
or not. Thank you!!

Signed-off-by: Fernand Galiana <fernand.galiana@isovalent.com>

gentoo-root commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -715,27 +718,42 @@ func isXfrmStateCilium(state netlink.XfrmState) bool {
 		}
 	}
+	if errs := ee.Errors(); errs != nil {
+		return errors.Join(errs...)
```
Why not use `ee.Error()`? Same below.

derailed replied:
Indeed! Now the initial impetus was to dedup state errors only. Not sure if policies can be deduped??
So per your comment, I've updated this PR to dedup both policies and states. Also noticed that there are several places where this actually occurs.
Please double check me on this! Thank you.

gentoo-root replied:
>Not sure if policies can be deduped??

Me neither, but I don't see why not.

sayboras reviewed (approved):
LGTM :heavy_check_mark:

gentoo-root commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -239,6 +243,11 @@ func xfrmStateReplace(new *netlink.XfrmState) error {
 		}
 	}
+	if err := errs.Error(); err != nil {
+		// TODO: Should this be a hard error and return instead??
```
If we return, neither the old state nor the new state will exist.

If we go on, though, they will conflict, and xfrmDeleteConflictingState will attempt to make one more attempt to remove the old state, which is likely to fail with the same error. So we'll end up in the same state, but we'll lose the original error. I guess it makes sense to return right here while we still have the proper error message.

derailed replied:
Thank you Max! Makes perfect sense
